### ISSUE TITLE
num_nodes in message() function

### ISSUE THREAD

**Issue by xcfcode** [acbull/pyHGT#3]
Thanks for this excellent job. 👍👍👍👍👍

In the following line, actually **data_size** is used as **num_nodes**.
https://github.com/acbull/pyHGT/blob/0d40a39e2505b29d7f044d3162cd7152f022aee8/conv.py#L57
and will be used in this the `softmax` function.
https://github.com/acbull/pyHGT/blob/0d40a39e2505b29d7f044d3162cd7152f022aee8/conv.py#L101
However, **data_size** indicates the number of edges, in which the edge index may larger than **data_size**, the **num_node** should be the max number in edge index + 1 as follows:
```
def maybe_num_nodes(index, num_nodes=None):
    return index.max().item() + 1 if num_nodes is None else num_nodes
```
https://github.com/rusty1s/pytorch_geometric/blob/378d4749559c0fb737375be61071c2cad2c50ae1/torch_geometric/utils/num_nodes.py#L2

When I try my own dataset, I modify the code as follows, which allows the torch geometric pkg to compute the num_nodes dynamically.

```        
self.att = softmax(res_att, edge_index_i)
```

Could you please give me some advice?

**Comment by acbull** [comment]
Hi:

You can check the official softmax function provided here: https://github.com/rusty1s/pytorch_geometric/blob/378d4749559c0fb737375be61071c2cad2c50ae1/torch_geometric/utils/softmax.py

which says:

```
num_nodes (int, optional): The number of nodes, *i.e.*
            :obj:`max_val + 1` of :attr:`index`. (default: :obj:`None`)
```

Also, the usage of this num_nodes is similar to setting an array with each node index as a key. Thus, you can either (1) omit num_nodes and let the softmax itself to check the bound by "maybe_num_nodes" function; or (2) use a larger value than the maximum value of the index.

For example:

```
from torch_geometric.utils import softmax
import torch

a = torch.randn(100)
i = torch.randint(0, 10, (100,))

x = softmax(a, i, num_nodes=10)
y = softmax(a, i, num_nodes=15)
z = softmax(a, i)
```

You can take a look, and see that x, y, and z are all the same.

But you cannot execute

```
softmax(a, i, num_nodes = 5)
```

As the index is larger than the num_nodes, and will raise overflow interruption.

**Comment by acbull** [comment]
Yes, you are right. This code is not that accurate, and I'll correct it.

But it won't influence the results actually, as long as data_size is larger than the max number of edge_index + 1. The reason is that the usage of this num_nodes is similar to setting an array with each node index as a key. Thus, you can either (1) omit num_nodes and let the softmax itself to check the bound by "maybe_num_nodes" function; or (2) use a larger value than the maximum value of the index.

For example:

```
from torch_geometric.utils import softmax
import torch

a = torch.randn(100)
i = torch.randint(0, 10, (100,))

x = softmax(a, i, num_nodes=10)
y = softmax(a, i, num_nodes=15)
z = softmax(a, i)
```

You can take a look, and see that x, y, and z are all the same.

But you cannot execute

```
softmax(a, i, num_nodes = 5)
```

As the index is larger than the num_nodes, and will raise overflow interruption.

Thanks for pointing this out. I'll simply remove the data_size term in the softmax for clarification.

**Comment by xcfcode** [comment]
Thanks a lot, since this repo is very useful for other tasks and the codes are well written. **omit num_nodes and let the softmax itself to check the bound by "maybe_num_nodes" function** will be more elegant. 
Anyway, thanks a lot for this beautiful repo !!!
